### ISSUE TITLE
[FEATURE REQUEST] Separate per-channel buffers

### ISSUE THREAD

**Issue by lgritz** [OpenImageIO/oiio#2744]
All of the basic ImageInput/ImageOutput functionality assumes that the in-memory buffer has "contiguous" pixels with interleaved channels (RGB RGB RGB...). It is currently impossible to read channels into a buffer with a "separate" configuration (RRRRR GGGGG BBBBB), nor to generally put each channel into a different buffer entirely. There is no combination of strides that will do this properly in a single read or write. You could read into a temporary contiguous buffer and then make nchannels calls to OIIO::convert_image (or parallel_convert_image) to copy each channel out into a separate permanent area.

That is certainly a relatively rare desire (evidenced by how few people have requested this over the years), but not an unreasonable one for some applications.

This issue is a placeholder for this work. I don't have time to tackle it right now, but I would welcome a PR that adds this capability in a clean way. Below I will sketch out how I think it would work and even break it into pieces which can be done one at a time.

- [ ] The first step is to nail down the user API by adding ImageInput/ImageOutput API calls declarations to imageio.h for this mode. I'll describe what I would recommend for ImageInput::read_scanlines, and let you extrapolate that into read_tiles, read_image, and the equivalent methods for ImageOutput.

     Note that there is a simple read_scanline() which reads only one scanline, all channels. And there is also a more complicated read_scanlines() that has many more parameters and has the ability to read any number of contiguous scanlines, as well as a range of channels which may be a contiguous subset of the total channel set. I recommend that we only add new versions of read_scanlines(), keep the simple read_scanline super simple. Same with read_tile vs read_tiles.

    So I would add a new method, `read_scanlines_separate()` which mimics read_scanlines, except that instead of parameter `void *data`, it takes `cspan<void*> data`.  The `data` span needs to have either size=1 or size = chend-chbegin, or it's an error. That is, a single data pointer that implies that the separate channel planes should be adjacent to each other in memory, or one data pointer for each channel which could point to totally arbitrary areas. Similarly, the `format` and `stride` parameters should also be cspan's, again where each one of which could take a single scalar value (meaning it's the same for each channel plane) but optionally allowing the span to contain (chend - chbegin) elements, giving potentially a different format or set of strides for each channel.

- [ ] Similarly, there will need to be an additional virtual method `read_native_scanlines_separate` which is like `read_native_scanlines` except with the data pointer now a cspan rather than a scalar. (Same for tiles.)

- [ ] In imageinput.cpp and imageoutput.cpp, the default implementation of the `read_native_*_separate` should be to call the non-separate one into a temporarily allocated buffer, then use OIIO::convert_image to do the individual per-channel conversions and copies into the multiple buffers that the caller requested. This is potentially inefficient -- equivalent to what people are doing now by reading interleaved pixels and then copying them out to separate buffers -- but it will be the fallback for those readers and writers that don't have a good way to read separate image planes efficiently (or which nobody bothers to subclass to add these methods).

- [ ] Also in imageinput.cpp and imageoutput.cpp, we'll need the full implementation of `read_scanlines_separate`. Start by copying `read_scanlines` and then, you know, do your thing.

- [ ] OpenEXR and TIFF: Now you can add overrides for `read_native_scanlines_separate` to the individual format readers. It should be very straightforward for OpenEXR, whose API naturally lends itself to this. It should be not too hard for TIFF, though you'll need to distinguish between reading files where it's already stored as "separate" planes (easy and efficient!) versus "contiguous" files, in which case you'll want to fall back to the parent class's default implementation that reads the contig and then copies the individual channels to the right spots. 

- [ ] Do a survey of all the other formats. Probably most of them don't have a way to store or read in "separate" configuration, so there's no advantage to overriding the default read_native_scanlines_separate. Some that do support it may be such unimportant formats (to us) that you elect not to do this extra work. You may decide that OpenEXR and TIFF are all we need, or you may find other formats where this is useful and straightforward to try to do this efficiently.

**Comment by JanHett** [comment]
Since I was the one who kicked this loose I thought I'd take a shot at this. However, while looking through the code, it occurred to me that reading images channel-wise can actually be expressed in the current API as repeated calls to `read_*` with arguments `chbegin` and `chend` of `0 <= chbegin < spec.nchannels` and `chend == chbegin + 1`.

I benchmarked this with a small JPG image as well as with a ~75MP 16bpc TIFF (which contains separate data) and the repeated calls are indeed faster (output from Google Benchmark):

```
----------------------------------------------------------------------------------
Benchmark                                        Time             CPU   Iterations
----------------------------------------------------------------------------------
BM_load_image_small_jpg_manual_scramble   30297767 ns     30281565 ns           23
BM_load_image_small_jpg_separate_passes   21570445 ns     21446375 ns           32
BM_load_image_75mp_tiff_manual_scramble 9079482621 ns   9073129000 ns            1
BM_load_image_75mp_tiff_separate_passes 3885153500 ns   3883243000 ns            1
```

Looking at the code, the relevant bit of the base class implementation [seems to be here](https://github.com/OpenImageIO/oiio/blob/b2b135284c1aad5702ed1771806d1ce0af641f18/src/libOpenImageIO/imageinput.cpp#L372) (for scanlines; AFAICT the [TIFF ImageInput just calls the base class implementation for this case](https://github.com/OpenImageIO/oiio/blob/b2b135284c1aad5702ed1771806d1ce0af641f18/src/tiff.imageio/tiffinput.cpp#L2028)).

I also wanted to benchmark different data layouts in the TIFF as well as doing the manual scramble with `convert_image` instead of my "dumbest thing that can work" approach of nested loops but I couldn't play around much more because reading the TIFF kept throwing a segfault upon calling `OIIO::ImageInput::open`. Not sure yet what's going on there.

Once I figure out what's causing the crash while reading TIFFs I'll experiment a bit more.

**Comment by lgritz** [comment]
You are correct that you could just make nchannels separate calls to read_scanlines/tiles/image, each requesting a single channel. That slipped my mind, too. 

How are you doing the manual scramble? Did you just write the loop yourself?

Note that [our implementation uses a parallel_for](https://github.com/OpenImageIO/oiio/blob/master/src/libOpenImageIO/imageinput.cpp#L395), so if you are writing a naive loop, the speed gain is simply that my version is multithreaded, and you have enough cores that this is making up for the fact that you're potentially doing the reads multiple times. And though you're calling it three times (say), each time it only needs to copy 1/3 of the data, and also each time you are doing any necessary format conversions on 1/3 of the data. So if the whole operation is dominated by the conversion time and memory bandwidth of the final copy to the user's buffer, the penalty for reading channels separately may be very low. 

THIS CHANGES EVERYTHING!

Forget everything I said in the description of this ticket. We can make this all work without adding any new user-level API calls at all.

I think the very next thing  you should do is also benchmark the "read all channels at once, in the default interleaved way" as a baseline. Maybe the penalty for separate reads of each channel so is so low that at this point we are done and there's nothing left to do: The standard advice we give people who want to read each channel to separate buffers should be to call read_scanlines/read_tiles/read_image once for each individual channel they need. And we can assure them that this is not N times less efficient.

ONLY if the penalty of separate-channel-reads versus one-channel-read is very high, I think the next thing to do is that for individual formats, you might want to add a custom read_native_scanlines/read_native_tiles in the variety that optimizes for the one-channel case by doing only the necessary work. (Overriding the default implementation you noticed above.)

TIFF could probably stand this, at least for the case of TIFF files that *store* the channels in separate planes (you really could do only a per-channel read in that case). Looks like OpenEXR already does this! So I think that **today**, making multiple calls to read_blah(), one per channel, is already as efficient as it's ever going to get, because that case is truly doing very close to the minimal work it would take to read the channels into separate buffers.

I lied just a teensy bit. Although it may turn out that the read case is already very efficient simply by using the per-channel read calls, there is no equivalent for write. The only way to write is to present a buffer containing all channels, interleaved. If it is also important that people be able to write from separated buffers and do so efficiently (i.e. without a "gather" step to assemble an interleaved buffer), then we may need to add some new write methods. But that can wait until later, probably, because I suspect that since most image files are read many more times than they are written, most apps consider the read time to be the more important to optimize.

**Comment by JanHett** [comment]
Yes, I was writing a naive loop myself, and I did suspect that that wasn't the fastest option. So, here are the results after a more thorough benchmarking session:

```
----------------------------------------------------------------------------------------------------------------
Benchmark                                                                      Time             CPU   Iterations
----------------------------------------------------------------------------------------------------------------
BM_read_image_75mp_tiff_c1_simple_read                                5885130225 ns   5882146000 ns            1
BM_read_image_75mp_tiff_c1_manual_scramble                           11223780206 ns  11217610000 ns            1
BM_read_image_75mp_tiff_c1_parallel_convert_image_scramble            9946731325 ns   9939592000 ns            1
BM_read_image_75mp_tiff_c1_convert_image_scramble                    10325258587 ns  10320673000 ns            1
BM_read_image_75mp_tiff_c1_separate_passes                            9089096504 ns   9079056000 ns            1

BM_read_image_75mp_tiff_ps_simple_read                                4337157980 ns   4332831000 ns            1
BM_read_image_75mp_tiff_ps_manual_scramble                            9968578300 ns   9962364000 ns            1
BM_read_image_75mp_tiff_ps_parallel_convert_image_scramble            9052325510 ns   9028931000 ns            1
BM_read_image_75mp_tiff_ps_convert_image_scramble                     9235743674 ns   9226922000 ns            1
BM_read_image_75mp_tiff_ps_separate_passes                            5697919061 ns   5619732000 ns            1

BM_read_image_75mp_tiff_separate_data_simple_read                     4077083009 ns   4074352000 ns            1
BM_read_image_75mp_tiff_separate_data_manual_scramble                 9427483713 ns   9423164000 ns            1
BM_read_image_75mp_tiff_separate_data_parallel_convert_image_scramble 8274093222 ns   8264988000 ns            1
BM_read_image_75mp_tiff_separate_data_convert_image_scramble          8623796187 ns   8619996000 ns            1
BM_read_image_75mp_tiff_separate_data_separate_passes                 4052297475 ns   4050250000 ns            1
```

TL;DR:

1) separate passes with `chbegin` and `chend` are always fastest
2) the speed-up depends a lot on how the file was produced: a TIFF coming from Capture One (C1) had barely any speed-up (and was slowest overall) while those produced by opening this original C1 TIFF in Photoshop (PS) and re-saving it showed a more pronounced difference
3) For PS TIFFs, reading in separate passes takes about the same time as just reading the file without any bells and whistles (i.e. just `file_ps->read_image(spice::helpers::type_to_typedesc<T>(), img_data.data());` and nothing else)
4) For PS TIFFs with separate channel data, the simple read and multiple passes take the same amount of time (sometimes one is faster, sometimes the other)

Sounds to me like this is already almost as good as it can get. Makes me wonder what C1 is doing to its TIFFs, though...

As for the write case... I agree that it isn't nearly as critical as reading, especially since it can be "hidden" from the user by doing it in the background. I'll still dig into the writing code a bit more to see if anything can be done there since I haven't looked at it very closely yet.

**Comment by lgritz** [comment]
It does indeed look like we're doing quite well. I think that you could speed up the separate_passes case for separate data by implementing the variety of TIFFInput::read_native_scanlines that takes a channel range, by doing the work for just the requested channel. But otherwise, we're doing quite well.

The tiff_c1 is curious.  Can you try `oiiotool -info -v` for the c1, ps, and separate inputs and report the full output? Maybe we can spot what the c1 is doing differently that's so inefficient.

**Comment by JanHett** [comment]
So, I could get the info for the C1 and interleaved PS files, but not the separate one (more on that later):

**Capture One generated:**
```
Reading data/testing/eibsee.tif
data/testing/eibsee.tif : 7466 x 9937, 3 channel, uint16 tiff
    2 subimages: 7466x9937 [u16,u16,u16], 120x160 [u8,u8,u8]
    channel list: R, G, B
    compression: "zip"
    Copyright: "Jan Hettenkofer"
    DateTime: "2020-06-26T10:59:27"
    ExposureTime: 0.01
    ICCProfile: 0, 0, 2, 48, 65, 68, 66, 69, 2, 16, 0, 0, 109, 110, 116, 114, ... [560 x uint8]
    Make: "Phase One"
    Model: "IQ180"
    Orientation: 1 (normal)
    PixelAspectRatio: 1
    planarconfig: "contig"
    ResolutionUnit: "in"
    Software: "Capture One 20 Macintosh"
    XResolution: 300
    YResolution: 300
    aux:Firmware: "IQ180, User Firmware: 8.00.30"
    aux:ImageNumber: "26413"
    aux:Lens: "-- mm f/16-2"
    aux:SerialNumber: "FP010055"
    Exif:DateTimeDigitized: "2020:06:26 10:59:27"
    Exif:DateTimeOriginal: "2020:06:26 10:59:27"
    Exif:ExposureBiasValue: 0
    Exif:ExposureProgram: 1 (manual)
    Exif:FocalPlaneResolutionUnit: 3 (cm)
    Exif:FocalPlaneXResolution: 1923.08
    Exif:FocalPlaneYResolution: 1923.08
    Exif:ImageUniqueID: "00005000007200000400E0580000672D"
    Exif:LightSource: 1 (daylight)
    Exif:MeteringMode: 3 (spot)
    Exif:SensingMethod: 2 (1-chip color area)
    Exif:ShutterSpeedValue: 6.64386 (1/99 s)
    Exif:WhiteBalance: 1 (manual)
    IPTC:CopyrightNotice: "Jan Hettenkofer"
    IPTC:DateCreated: "20200626"
    IPTC:DigitalCreationDate: "20200626"
    IPTC:DigitalCreationTime: "105927"
    IPTC:TimeCreated: "105927"
    oiio:BitsPerSample: 16
    oiio:ColorSpace: "sRGB"
    photoshop:LegacyIPTCDigest: "99401C8F139D4A438150893C3DC475A9"
    tiff:Compression: 32946
    tiff:PhotometricInterpretation: 2
    tiff:PlanarConfiguration: 1
    tiff:RowsPerStrip: 176
```

**Saved with PS (interleaved)**
```
Reading data/testing/eibsee_ps.tif
data/testing/eibsee_ps.tif : 7466 x 9937, 3 channel, uint16 tiff
    channel list: R, G, B
    compression: "zip"
    Copyright: "Jan Hettenkofer"
    DateTime: "2020-06-26T10:59:27"
    ExposureTime: 0.01
    ICCProfile: 0, 0, 2, 48, 65, 68, 66, 69, 2, 16, 0, 0, 109, 110, 116, 114, ... [560 x uint8]
    Make: "Phase One"
    Model: "IQ180"
    Orientation: 1 (normal)
    PixelAspectRatio: 1
    planarconfig: "contig"
    ResolutionUnit: "in"
    Software: "Capture One 20 Macintosh"
    XResolution: 300
    YResolution: 300
    aux:Firmware: "IQ180, User Firmware: 8.00.30"
    aux:ImageNumber: "26413"
    aux:Lens: "-- mm f/16-2"
    aux:SerialNumber: "FP010055"
    Exif:ColorSpace: 65535
    Exif:DateTimeDigitized: "2020:06:26 10:59:27"
    Exif:DateTimeOriginal: "2020:06:26 10:59:27"
    Exif:ExposureBiasValue: 0
    Exif:ExposureProgram: 1 (manual)
    Exif:FocalPlaneResolutionUnit: 3 (cm)
    Exif:FocalPlaneXResolution: 1923.08
    Exif:FocalPlaneYResolution: 1923.08
    Exif:ImageUniqueID: "00005000007200000400E0580000672D"
    Exif:LightSource: 1 (daylight)
    Exif:MeteringMode: 3 (spot)
    Exif:PixelXDimension: 7466
    Exif:PixelYDimension: 9937
    Exif:SensingMethod: 2 (1-chip color area)
    Exif:ShutterSpeedValue: 6.64386 (1/99 s)
    Exif:WhiteBalance: 1 (manual)
    IPTC:DocumentID: "adobe:docid:photoshop:e1e1220d-1af0-cc4a-96f7-b9f2659f91b6"
    IPTC:InstanceID: "xmp.iid:889c2a2e-de27-4287-a3fc-6a960364ca9c"
    IPTC:MetadataDate: "2020-10-17T14:29:07+02:00"
    IPTC:ModifyDate: "2020-10-17T14:29:07+02:00"
    IPTC:OriginalDocumentID: "xmp.did:704c8887-7333-40b5-b05c-7aaca8824f06"
    oiio:BitsPerSample: 16
    photoshop:ColorMode: "3"
    photoshop:ICCProfile: "Adobe RGB (1998)"
    photoshop:LegacyIPTCDigest: "99401C8F139D4A438150893C3DC475A9"
    rdf:parseType: "Resource"
    stEvt:action: "saved"
    stEvt:changed: "/"
    stEvt:instanceID: "xmp.iid:889c2a2e-de27-4287-a3fc-6a960364ca9c"
    stEvt:softwareAgent: "Adobe Photoshop 21.2 (Macintosh)"
    stEvt:when: "2020-10-17T14:29:07+02:00"
    tiff:Compression: 8
    tiff:PhotometricInterpretation: 2
    tiff:PlanarConfiguration: 1
    tiff:RowsPerStrip: 23
    tiff:subfiletype: 0
```

**Saved with PS (separate)**

`oiiotool -info -v data/testing/eibsee_separate.tif` crashes with the following stack trace:

```
0# OpenImageIO_v2_1::Sysutil::stacktrace() in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
 1# OpenImageIO_v2_1::stacktrace_signal_handler(int) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
 2# _sigtramp in /usr/lib/system/libsystem_platform.dylib
 3# OpenImageIO_v2_1::ParamValueList::find(OpenImageIO_v2_1::string_view, OpenImageIO_v2_1::TypeDesc, bool) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
 4# OpenImageIO_v2_1::ImageSpec::attribute(OpenImageIO_v2_1::string_view, OpenImageIO_v2_1::TypeDesc, void const*) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
 5# OpenImageIO_v2_1::ImageSpec::attribute(OpenImageIO_v2_1::string_view, OpenImageIO_v2_1::string_view) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
 6# OpenImageIO_v2_1::TIFFInput::get_string_attribute(OpenImageIO_v2_1::string_view, int) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
 7# OpenImageIO_v2_1::TIFFInput::find_tag(int, TIFFDataType, OpenImageIO_v2_1::string_view) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
 8# OpenImageIO_v2_1::TIFFInput::readspec(bool) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
 9# OpenImageIO_v2_1::TIFFInput::seek_subimage(int, int) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
10# OpenImageIO_v2_1::TIFFInput::open(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, OpenImageIO_v2_1::ImageSpec&) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
11# OpenImageIO_v2_1::TIFFInput::open(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, OpenImageIO_v2_1::ImageSpec&, OpenImageIO_v2_1::ImageSpec const&) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
12# OpenImageIO_v2_1::pvt::ImageCacheFile::open(OpenImageIO_v2_1::pvt::ImageCachePerThreadInfo*) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
13# OpenImageIO_v2_1::pvt::ImageCacheImpl::verify_file(OpenImageIO_v2_1::pvt::ImageCacheFile*, OpenImageIO_v2_1::pvt::ImageCachePerThreadInfo*, bool) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
14# OpenImageIO_v2_1::pvt::ImageCacheImpl::get_image_info(OpenImageIO_v2_1::pvt::ImageCacheFile*, OpenImageIO_v2_1::pvt::ImageCachePerThreadInfo*, int, int, OpenImageIO_v2_1::ustring, OpenImageIO_v2_1::TypeDesc, void*) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
15# OpenImageIO_v2_1::pvt::ImageCacheImpl::get_image_info(OpenImageIO_v2_1::ustring, int, int, OpenImageIO_v2_1::ustring, OpenImageIO_v2_1::TypeDesc, void*) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
16# input_file(int, char const**) in /usr/local/bin/oiiotool
17# OpenImageIO_v2_1::ArgParse::Impl::parse(int, char const**) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
18# getargs(int, char**) in /usr/local/bin/oiiotool
19# main in /usr/local/bin/oiiotool
 0# OpenImageIO_v2_1::Sysutil::stacktrace() in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
 1# OpenImageIO_v2_1::stacktrace_signal_handler(int) in /usr/local/Cellar/openimageio/2.1.18_1/lib/libOpenImageIO.2.1.18.dylib
 2# _sigtramp in /usr/lib/system/libsystem_platform.dylib
[1]    2677 abort      oiiotool -info -v data/testing/eibsee_separate.tif
```

This seems to be related to the flaky segfault I was running into during my brenchmarking experiments (the stack-trace looks really similar) except that in this case it always happens. Might be something wrong with my installation, though. I'll see if I can reproduce this with a built-from-source version, too.

**Comment by lgritz** [comment]
Do you think you could privately email me the file that frequently crashes?

**Comment by lgritz** [comment]
There's something really subtle and interesting here. I think my bug. Investigating.

**Comment by lgritz** [comment]
#2757 proposes a partial fix (for all the common cases) for the crashing and incorrect reads. And maybe it also improves performance when reading planarconfig=separate TIFF files.

**Comment by lgritz** [comment]
The really fast summary of #2757 is that there was a bug when reading planarconfig=separate TIFF files that used rowsperstrip > 1 in combination with certain compression codecs. OIIO's TIFF output forced rps=1 when writing 'separate' files, so in practice we never saw the symptoms, which could only be observed when using OIIO to read TIFF files written by non-OIIO-based apps (and only with the offending combination of file properties).

**Comment by lgritz** [comment]
Now that you've added docs explaining how to do read into separate buffers, and we have concluded that it's not especially inefficient to do so by this method, is this issue ready to close?

**Comment by JanHett** [comment]
My use case is covered by the fix and the additional docs. Feel free to close the issue :).
